File: leetcode/python/1046_last_stone_weight.py
import heapq


class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        heap = [-n for n in stones]
        heapq.heapify(heap)
        while len(heap) > 1:
            x = -heapq.heappop(heap)
            y = -heapq.heappop(heap)
            if x - y > 0:
                heapq.heappush(heap, y-x)
        return -heap[0] if heap else 0

s = Solution()
print(s.lastStoneWeight([2,7,4,1,8,1]))
print(s.lastStoneWeight([1]))

File: leetcode/c++/1046_last_stone_weight.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        make_heap(stones.begin(), stones.end());
        while (stones.size() > 1) {
            pop_heap(stones.begin(), stones.end());
            int y = stones.back();
            stones.pop_back();
            pop_heap(stones.begin(), stones.end());
            int x = stones.back();
            stones.pop_back();
            if (x != y) {
                stones.push_back(y - x);
                push_heap(stones.begin(), stones.end());
            }
        }
        return stones.empty() ? 0 : stones.front();
    }
};